src/Common/src/TypeSystem/Common/TypeSystemConstraintsHelpers.cs

```diff
@@ -2,13 +2,35 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using System.Diagnostics;
 
 namespace Internal.TypeSystem
 {
+    public class InstantiationContext
```

Comment:
Nit: The style in the TypeSystem is to just use readonly fields for situations like these.

```
public class InstantiationContext
{
    public readonly Instantiation TypeInstantiation;
    public readonly Instantiation MethodInstantiation;

    public InstantiationContext(Instantiation typeInstantiation, Instantiation methodInstantiation)
    {
        TypeInstantiation = typeInstantiation;
        MethodInstantiation = methodInstantiation;
    }
}
```